spec/services/deliver_spec.rb

```diff
@@ -65,6 +80,68 @@
       end
     end
 
+    shared_examples 'a delivery failure' do |options|
+      options ||= {}
+      
+      it "raises a Routemaster::Exceptions::DeliveryFailure exception" do
+        expect { perform }.to raise_error(Routemaster::Exceptions::DeliveryFailure)
+      end
+
+      it_behaves_like 'an event counter', options.merge(count: 3, status: 'failure')
+      it_behaves_like 'it updates the HP of the subscriber', by: -2
+    end
+
+    shared_examples 'it updates the HP of the subscriber' do |options|
+      by = options.fetch(:by)
+      description = (by >= 0) ? 'increases' : 'decreases'
+
+      it "#{description} the health points of the subscriber by #{by}" do
+        expect { perform rescue nil }.to change {
+          reloaded_subscriber.health_points
+        }.from(base_hp).to(base_hp + by)
+      end
+    end
+
+    shared_examples_for 'a subscriber throttler' do
+      let(:throttle) do
+        instance_double(Routemaster::Services::Throttle, notice_success: base_hp, notice_failure: base_hp)
+      end
+
+      before do
+        allow(Routemaster::Services::Throttle).to receive(:new).with(subscriber).and_return(throttle)
```

Comment:
I've avoided stubbing throughout the codebase so far — mainly because it avoids weird testing bugs.
Could I propose, instead of stubbing the constructor, to support dependency injection here?

You could pass a `throttle_using: Routemaster::Services::Throttle` to the constructor, and use that to inject your test double.